keps/sig-storage/20200120-skip-permission-change.md

```diff
@@ -0,0 +1,166 @@
+allow the user to specify whether they want the permission and ownership change to occur for that volume.
+
+## Motivation
+
+When a volume is mounted on the node, we recursively change permissions of volume
+before bind mounting the volume inside container. The reason of doing this is to ensure
+that volumes are readable/writable by provided fsGroup.
+
+But this presents following problems:
+ - An application(many popular databases) which is sensitive to permission bits changing
+   underneath may refuse to start whenever volume being used inside pod gets mounted on
+   different node.
+ - If volume has a large number of files, performing recursive `chown` and `chmod`
+   could be slow and could cause timeout while starting the pod.
+
+### Goals
+
+ - Allow volume ownership and permission change to be skipped during mount
+
+### Non-Goals
+
+ - In some cases if user brings in a large enough volume from outside, the first time ownership and permission change still could take lot of time.
+ - On SELinux enabled distributions we will still do recursive chcon whenever applicable and handling that is outside the scope.
+ - This proposal does not attempt to fix two pods using same volume with conflicting fsgroup. `FSGroupPermissionChangePolicy` also will be only applicable to volume types which support setting fsgroup.
+
+## Proposal
+
+We propose that an user can optionally opt-in to skip recursive ownership(and permission) change on a volume if volume already has right permissions.
+
+### Implementation Details/Notes/Constraints [optional]
+
+Currently Volume permission and ownership change is done using a breadth-first algorithm starting from root directory of the volume. As part of this proposal we will change the algorithm to modify permissions/ownership of the root directory after all directories/files underneath have their permissions changed.
+
+When creating a pod, we propose that `pod.Spec.Volumes[]` field expanded to include a new field called `FSGroupPermissionChangePolicy` which can have following possible values:
+
+ - `Always` --> Always change the permissions and ownership to match fsGroup. This is the current behavior and it will be the default one when this proposal is implemented. Algorithm that performs permission change however will be changed to perform permission change of top level directory last.
+ - `OnRootDirPermissionMismatch` --> Only perform permission and ownership change if permissions of top level directory does not match with expected permissions and ownership.
+
+```go
+type PodFSGroupPermissionChangePolicy string
+
+const(
+    ChangePermissionOnRootDirmismatch PodFSGroupPermissionChangePolicy = "OnRootDirPermissionMismatch"
+    AlwaysChangeVolumePermission PodFSGroupPermissionChangePolicy = "Always"
+)
+
+// Volume represents a named volume in a pod that may be accessed by any containers in the pod.
+type Volume struct {
```

Comment:
Another option is to add this FSGroupPermissionChangePolicy option to Pod.Spec.SecurityContext where the fsGroup is set.

Currently fsGroup will be applied to all volumes that are used by this pod since this is pod level option. Is there a use case that different volumes need different policies?

Reply:
done.

Reply:
Ha, I suggested moving it out of Pod.Spec.SecurityContext :)

IIUC, the primary motivation for this is to be applied to very large volumes, when the permissions change can be slow. If a pod mounts multiple small volumes in addition to the large one, it may make sense to only apply the option to the large volume.